### PR TITLE
[fix](util) Fix DorisOnce not accurate once when exception happend

### PR DESCRIPTION
**Pull request by Hastyshell**
## Proposed changes

In cpp, exception catch relies on stack unwind and will destruct all stack variables before entering catch block.

In the original DorisOnce, if exception happended when `fn` called, the `_flag_lock` will be released before `_has_called` flag in set. The race condition may make it actually called twice.

This PR use std::once_falg and std::call_once to refactor the DorisOnce.

1. Support call once semantic and return result when `call()` is called. If an exception thrown, the exception will be rethrown whenever called the `call()`.
2. Supports to get is called status through function `is_called()`.
3. Support to get produced value through `stored_result()`. If exception throwed, will rethrow this exception. If it is not called, will throw an exception to indicate it.

### CODE REVIEW

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

File: be/src/util/once.h [hunk 1]
```diff
@@ -89,16 +89,15 @@ class DorisCallOnce {
             return _status;
         }
         try {
+            // This memory order make sure both status and eptr is set
+            // and will be seen in another thread.
+            _once_flag.store(true, std::memory_order_release);
             _status = fn();
         } catch (...) {
             // Save the exception for next call.
             _eptr = std::current_exception();
```
Comment (TangSiyang2001) [inline be/src/util/once.h:98]: how shall we handle exception here to make it visible intime for other threads?

Comment (TangSiyang2001) [reply be/src/util/once.h:98]: I got a good idea, will update this PR later.

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run p0

File: be/src/util/once.h [hunk 2]
```diff
@@ -42,95 +48,67 @@ namespace doris {
...
-    // occurred.
     template <typename Fn>
-    ReturnType call(Fn fn) {
-        // Avoid lock to improve performance
-        if (has_called()) {
-            if (_eptr) {
-                std::rethrow_exception(_eptr);
+        requires std::is_invocable_r_v<R, Fn>
+    R call(Fn fn) {
+        std::call_once(_once_flag, [&] {
```
Comment (yiguolei) [inline be/src/util/once.h:66]: In the past, we use std call once, but there are 2 main reasons that we implement our own call once:
1. std call once could not deal with exception logic correctly.
2. when it is core during std call once, it's callstack will include another thread, it is not very clear.

**Review by yiguolei — CHANGES_REQUESTED** [review]
https://learn.microsoft.com/en-us/cpp/cpp/exceptions-and-stack-unwinding-in-cpp?view=msvc-170

Are you sure the lock will be destroyed?

**Comment by TangSiyang2001** [comment]
The call once `called twice` problem is in the formal version DorisOnce, the current version has no problem, this fix is not in demand, so close this PR.

**Comment by TangSiyang2001** [comment]
> https://learn.microsoft.com/en-us/cpp/cpp/exceptions-and-stack-unwinding-in-cpp?view=msvc-170
> 
> Are you sure the lock will be destroyed?

Misdirect by a similar case, the lock is out of try{}, and will not be destroyed.

**Pull request closed without merging**